launch/launch/launch_service.py

```diff
@@ -338,28 +338,34 @@ def _on_exception(loop, context):
                         ret = await self._shutdown(reason='idle', due_to_sigint=False)
                         assert ret is None, ret
                         continue
+
+                    # Stop running if we're shutting down and there's no more work
+                    if self.__shutting_down and is_idle:
+                        break
+
+                    # Collect futures to wait on
+                    entity_futures = [pair[1] for pair in self._entity_future_pairs]
+                    entity_futures.extend(self.__context._completion_futures)
+
+                    # Create a task to process any events in the queue
+                    # and add it to the list of awaitables
                     process_one_event_task = this_loop.create_task(self._process_one_event())
-                    if self.__shutting_down:
-                        # If shutting down and idle then we're done.
-                        if is_idle:
+                    entity_futures.append(process_one_event_task)
+
+                    # Wait on events and futures until an event is processed or a timeout occurs
+                    while True:
+                        done, pending = await asyncio.wait(
+                            entity_futures,
+                            timeout=1.0,
+                            return_when=asyncio.FIRST_COMPLETED
+                        )
+                        if process_one_event_task in done:
+                            break
+                        elif len(done) == 0:
                             process_one_event_task.cancel()
```

Comment:
why cancelling here? this was called only when shutdown happened

Reply:
It's possible that the `process_one_event_task` doesn't complete (timeout). Every time this happens, tasks were accumulating and during shutdown I would see all kinds of errors saying "task was destroyed but it is pending". Calling cancel here fixed those errors.

Before, we blocked forever until this task was completed so we didn't need to cancel.

Reply:
> It's possible that the process_one_event_task doesn't complete (timeout). Every time this happens, tasks were accumulating and during shutdown I would see all kinds of errors saying "task was destroyed but it is pending". Calling cancel here fixed those errors.

That is because `process_one_event_task` have to be cancelled if shutdown happened, but it shouldn't be cancelled just because nothing was done. We actually don't strictly need a timeout in the above wait.

Reply:
`process_one_event_task` goes out of scope if we timeout, so we can't cancel it later during shutdown unless we hold a reference to it (maybe there's a way to get a reference from asynchio?). If I remove the timeout case:

```diff
- elif len(done) == 0:
-    process_one_event_task.cancel()
-    break
```

Then my tests block forever again. I think this is because all futures complete, but there are no events in the queue.

Reply:
> Then my tests block forever again. I think this is because all futures complete, but there are no events in the queue.

Yes, the idea is that you check if `is_idle` again in the loop, and `break` it if that happens.

Reply:
We can check `is_idle` again in this loop, then we have to re-collect the entity futures. It seems a bit redundant. Instead, we could hold a reference to `process_one_event_task` and only create a new one if it is done. I guess the code would be a little bit simpler. But, I don't see the issue with canceling, unless you think it's an expensive operation.

Reply:
We do want to avoid accidentally canceling the processing mid-event. See 9c4c641 for a fix.